src/garbage_collector.rs

```diff
@@ -0,0 +1,208 @@
+use crate::config::GarbageCollectorOptions;
+use crate::error::SlateDBError;
+use crate::garbage_collector::GarbageCollectorMessage::*;
+use crate::manifest_store::{ManifestStore, StoredManifest};
+use crate::metrics::DbStats;
+use crate::tablestore::TableStore;
+
+enum GarbageCollectorMessage {
+    Shutdown,
+}
+
+pub(crate) struct GarbageCollector {
+    main_tx: crossbeam_channel::Sender<GarbageCollectorMessage>,
+    main_thread: Option<JoinHandle<()>>,
+}
+
+impl GarbageCollector {
+    pub(crate) async fn new(
+        manifest_store: Arc<ManifestStore>,
+        table_store: Arc<TableStore>,
+        options: GarbageCollectorOptions,
+        tokio_handle: Handle,
+        db_stats: Arc<DbStats>,
+    ) -> Result<Self, SlateDBError> {
+        let (external_tx, external_rx) = crossbeam_channel::unbounded();
+        let (err_tx, err_rx) = tokio::sync::oneshot::channel();
+        let tokio_handle = options.compaction_runtime.clone().unwrap_or(tokio_handle);
+        let main_thread = thread::spawn(move || {
+            let load_result = GarbageCollectorOrchestrator::new(
+                manifest_store,
+                table_store,
+                options,
+                tokio_handle,
+                external_rx,
+                db_stats,
+            );
+            let mut orchestrator = match load_result {
+                Ok(orchestrator) => orchestrator,
+                Err(err) => {
+                    err_tx.send(Err(err)).expect("err channel failure");
+                    return;
+                }
+            };
+            err_tx.send(Ok(())).expect("err channel failure");
+            orchestrator.run();
+        });
+        err_rx.await.expect("err channel failure")?;
+        Ok(Self {
+            main_thread: Some(main_thread),
+            main_tx: external_tx,
+        })
+    }
+
+    /// Close the garbage collector
+    pub(crate) async fn close(mut self) {
+        if let Some(main_thread) = self.main_thread.take() {
+            self.main_tx.send(Shutdown).expect("main tx disconnected");
+            main_thread
+                .join()
+                .expect("failed to stop main compactor thread");
+        }
+    }
+}
+
+struct GarbageCollectorOrchestrator {
+    manifest_store: Arc<ManifestStore>,
+    table_store: Arc<TableStore>,
+    options: GarbageCollectorOptions,
+    stored_manifest: StoredManifest,
+    tokio_handle: Handle,
+    external_rx: crossbeam_channel::Receiver<GarbageCollectorMessage>,
+    db_stats: Arc<DbStats>,
+}
+
+impl GarbageCollectorOrchestrator {
+    fn new(
+        manifest_store: Arc<ManifestStore>,
+        table_store: Arc<TableStore>,
+        options: GarbageCollectorOptions,
+        tokio_handle: Handle,
+        external_rx: crossbeam_channel::Receiver<GarbageCollectorMessage>,
+        db_stats: Arc<DbStats>,
+    ) -> Result<Self, SlateDBError> {
+        let stored_manifest =
+            tokio_handle.block_on(StoredManifest::load(manifest_store.clone()))?;
```

Comment:
I think this is a bit fragile, because it depends on how the tokio runtime is configured.

My usual approach here would be to make `new` itself `async`. I realize the codebase uses `block_on` in other places, so this is not a new limitation, but figured I'd mention it.

Reply:
+1, I remember there were many issues caused by `block_on` in Databend, leading to many hanging problems that were very hard to diagnose.

i can't recall the details of what caused these hangs, but the calling flow was something like this:

1. `block_on` lets a sync function await an `async` task to finish.
2. then, this sync function is called by another `async` function, and the compiler doesn't stop you from doing this.
3. the `async` function you called in `block_on` gets blocked by something, which might be resumed by another task, but this task's scheduled in a process calls this `sync` function, and it not awares about polling the other tasks.
4. since async/await in Rust is polling-based, now no one is driving the poll, so everything hangs

i guess `block_on` is not always considered as evil, but the oreo-style function calls `async` --> `sync` -(with block_on)-> `async` is not bad to get avoided when possible in practise.

Reply:
SGTM. I'm going to open a separate issue for this one. I think at least the compactor and GC would need to be changed.

One question, though: a nice thing about `block_on` is that we can allow users to supply separate async runtime handles to remove thread contention between the writer operations, the GC operations, and the compactor operations. Is there a way to achieve this without using the `block_on` method?

Reply:
i once came across this article from influxdb and i feel this approach makes some senses to seperate runtimes:
https://www.influxdata.com/blog/using-rustlangs-async-tokio-runtime-for-cpu-bound-tasks/

they used a `oneshot` channel to let the async runtimes communicate with each other, as `oneshot` channels are designed to be sync:

> Since the send method is not async, it can be used anywhere. This includes sending between two runtimes, and using it from non-async code.

and it seems that they also introduced a `block_on` to seperate the async runtime, the good part of this `block_on` is that it's only used to seperate the async runtime, and happens only once.

Reply:
We use `block_on` a bunch in the compactor to call async code from the compactor thread. I think we are not susceptible to this type of hang from there because that code can't really be called from an async context as it's all internal to the compactor, which sets up its own orchestrator thread. Does that seem reasonable @flaneur2020?

Also, from what I've seen tokio itself will panic with `Cannot start a runtime from within a runtime. This happens because a function (like block_on) attempted to block the current thread while the thread is being used to drive asynchronous tasks.` if you try to use `block_on` from async code running in a tokio runtime.

Tangentially, my main reason for writing the compactor as a background thread rather than a task was that I found synchronous code easier to write and debug, so I stuck to using synchronous code for processes that are fully owned by slatedb (like the compactor, as opposed to the main api which is called by code we don't control):
- The debugger is kind of annoying to use with async because from you can't step over `await` calls (the workaround I've been using is to just stick a breakpoint after the `await`, but that's kind of annoying).
- There's been times when I wanted to get a stack trace of the various things that are running, but I haven't seen any great tools for doing this for async code. I have been meaning to try https://github.com/tokio-rs/console which seems to be able to show you the tasks that are running at least, along with some diagnostic info about them.
- There's some constraints with async that are sort of annoying. For example, I all the data owned by the task has to be Send if the task does any awaits. But these are generally not a huge deal. The bigger one was the limitations around async traits, but then I learned about `async_trait`.

I'm definitely a newb to rust, so definitely would like any tips/pointers in this area, and whether this has come up for you guys (async being a pain to debug).

> I think this is a bit fragile, because it depends on how the tokio runtime is configured.

@paulgb is the concern here that if the handle is for a current_thread runtime, that block_on will hang if the runtime itself isn't in `block_on`? I suppose another option then is that we can check that using the `runtime_flavor` method and use our own runtime if that's the case.

Reply:
> We use block_on a bunch in the compactor to call async code from the compactor thread. I think we are not susceptible to this type of hang from there because that code can't really be called from an async context as it's all internal to the compactor, which sets up its own orchestrator thread. Does that seem reasonable @flaneur2020?

agree, i think this is the reasonable use case to start up a new runtime by using thread + `block_on`.

Reply:
> whether this has come up for you guys (async being a pain to debug).

to be honest, i hate the rust async ecosystem 🥲, there're simply too many sufferings around it. not to say the type system pains around it, it's also the only polling-driven async system in my knowledge, the execution flow of the code is not very intuitive at all. 🥲

however, in my understanding if one choose using `async`, the tokio ecosystem & practises are considered as the least painful one. (and surely there're plenty of people in the tokio community shares the pains with you)

on the case of backtrace, you could consider using `async_backtrace` to add backtraces to it (have to manually attach `#[async_backtrace]` to every function though. this is used in Databend after some failed battles with gdb stacks...)

Reply:
> to be honest, i hate the rust async ecosystem 🥲

Endorse.

Reply:
> We use `block_on` a bunch in the compactor to call async code from the compactor thread. I think we are not susceptible to this type of hang from there because that code can't really be called from an async context as it's all internal to the compactor, which sets up its own orchestrator thread. Does that seem reasonable @flaneur2020?

I see, it sounds like it is not dependent on the async runtime of the application code that calls into SlateDB then? If it's essentially just an implementation detail that doesn't interact with the application's tokio runtime, I think that's fine.

Reply:
Using a dedicated/background thread is the way to run merge/compaction and the problem here is using runtimes like tokio which schedules your thread by its own. Similarly in Lucene.NET we use merge threads for compaction and also we have the taskschduler version of the merge thread.

Dedicated Thread impl: https://github.com/apache/lucenenet/blob/0d7bba7a42a89a3e348c6c27370425b4db33b492/src/Lucene.Net/Index/ConcurrentMergeScheduler.cs#L570

TaskSchuduler impl: https://github.com/apache/lucenenet/blob/0d7bba7a42a89a3e348c6c27370425b4db33b492/src/Lucene.Net/Support/Index/TaskMergeScheduler.cs#L417


Thanks!